### ISSUE TITLE
Argument 1 of "__mul__" is incompatible with supertype "tuple" ; supertype defines the argument type as "SupportsIndex"

### ISSUE THREAD

**Issue by mortoray** [python/mypy#12740]
**Bug Report**

I have a NamedTuple that implements a _mul_ operator, to operator as a 2d vector.  MyPy is flagging the __mul__ function as invalid.

**To Reproduce**

Run `mypy` on this code:

```
from typing import NamedTuple

class Point2(NamedTuple):
	x: float
	y: float
	
	def __add__(self, o: 'Point2'): # type: ignore
		return Point2(self.x+o.x, self.y+o.y)

	def __mul__(self, o: float):
		return Point2(self.x*o, self.y*o)
```

**Expected Behavior**

I would expect this to type check correctly. I understand the Liskov substitution principle, but unsure how it applies here. I'm unaware that the NamedTuple base implements `__mul__` (the Python interpreter doesn't think it does). I don't understand how the "SupportsIndex" would prevent implementing a `__mul__` override.

**Actual Behavior**
You get this error:

```
/tmp/sample.py:10: error: Argument 1 of "__mul__" is incompatible with supertype "tuple"; supertype defines the argument type as "SupportsIndex"
/tmp/sample.py:10: note: This violates the Liskov substitution principle
/tmp/sample.py:10: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

Despite this error, mypy will apply type-checking against this mul function,correctly identifying problems elsewhere in the code, such as multiplying Point2 by Point2.

**Your Environment**

mypy 0.950 (compiled: yes)
python3.10

**Comment by erictraut** [comment]
I think mypy is correct in reporting this as an error. `NamedTuple` derives from `tuple`, and `tuple` does implement `__mul__`. The parameter type for `tuple.__mul__` is `SupportsIndex`. You're overriding this method with a version of `__mul__` that accepts a `float`, and `float` isn't type compatible with the protocol `SupportsIndex`.

Here's an example of how you run into trouble by violating the LSP:

```python
def func(x: tuple[float, float]) -> tuple[float, float, float, float]:
    return x * 2

p = Point2(0, 0)
v = func(p)

# The revealed type is inconsistent with the runtime type!
reveal_type(v) # tuple[float, float, float, float]
print(v) # Point2
```

**Comment by AlexWaygood** [comment]
> I'm unaware that the `NamedTuple` base implements `__mul__` (the Python interpreter doesn't think it does)

In addition to Eric's excellent points above, I think there's a slight misunderstanding here. `NamedTuple` isn't actually a class: it's a function. (Yup, it's possible to inherit from functions in very specific circumstances.) When you create a class by "inheriting" from `NamedTuple`, the resulting class doesn't actually have `NamedTuple` in its mro, because `NamedTuple` isn't a class. Instead, you should understand inheriting from `NamedTuple` for what it is: syntactic sugar for easily creating a custom `tuple` subclass that has a bunch of properties added to it (one for each field of the tuple).

**Comment by AlexWaygood** [comment]
This is the magic line in the CPython source code that means `NamedTuple` can be used as a base, by the way: https://github.com/python/cpython/blob/0924b95f6e678beaf4a059d679515956bac608fb/Lib/typing.py#L2860
